src/mapreduce.jl

```diff
@@ -0,0 +1,276 @@
+end
+
+@inline function Base.setindex!(o::_RollingCutOut, val, i)
+    @boundscheck checkbounds(o, i)
+    @inbounds o.array[i] = val
+end
+
+############################## tmap, tmap! ##############################
+struct MapWorkspace{F,B,AD,AS}
+    f::F
+    batches::B
+    arena_dst_view::AD
+    arena_src_views::AS
+end
+
+@noinline function run!(o::MapWorkspace)
+    let o=o
+        Threads.@threads for i in 1:length(o.batches)
+            tid = Threads.threadid()
+            dst_view  = o.arena_dst_view[tid]
+            src_views = o.arena_src_views[tid]
+            inds = o.batches[i]
+            dst_view.eachindex = inds
+            for view in src_views
+                view.eachindex = inds
+            end
+            Base.map!(o.f, dst_view, src_views...)
+        end
+    end
+end
+
+function create_arena_src_views(srcs, sample_inds)
+    nt = Threads.nthreads()
+    [Base.map(src -> _RollingCutOut(src, sample_inds), srcs) for _ in 1:nt]
+end
+
+@noinline function prepare(::typeof(tmap!), f, dst, srcs; batch_size::Int)
+    # we use IndexLinear since _RollingCutOut implementation
+    # does not support other indexing well
+    all_inds  = eachindex(IndexLinear(), srcs...)
+    batches   = Batches(all_inds, batch_size)
+    sample_inds = batches[1]
+    nt = Threads.nthreads()
+    arena_dst_view  = [_RollingCutOut(dst, sample_inds) for _ in 1:nt]
+    arena_src_views = create_arena_src_views(srcs, sample_inds)
+    return MapWorkspace(f, batches, arena_dst_view, arena_src_views)
+end
+
+@noinline function tmap!(f, dst, srcs::AbstractArray...;
+                        batch_size=default_batch_size(length(dst)))
+    isempty(first(srcs)) && return dst
+    w = prepare(tmap!, f, dst, srcs, batch_size=batch_size)
+    run!(w)
+    dst
+end
+
+function tmap(f, srcs::AbstractArray...; 
+             batch_size=default_batch_size(length(first(srcs)))
+            )
+    g = Base.Generator(f,srcs...)
+    T = Base.@default_eltype(g)
+    dst = similar(first(srcs), T)
+    tmap!(f, dst, srcs...; batch_size=batch_size)
+end
+
+############################## tmapreduce(like) ##############################
+struct Reduction{O}
+    op::O
+end
+(red::Reduction)(f, srcs...) = Base.mapreduce(f, red.op, srcs...)
+
+struct MapReduceWorkspace{R,F,B,V,OA<:OffsetArray}
+    reduction::R
+    f::F
+    batches::Batches{B}
+    arena_src_views::V
+    batch_reductions::OA
+end
+
+struct NoInit end
+
+function create_reduction(::typeof(tmapreduce), op)
+    Reduction(op)
+end
+
+function prepare(::typeof(tmapreduce), f, op, srcs; init, batch_size::Int)
+    red = create_reduction(tmapreduce, op)
+    w = prepare_mapreduce_like(red, f, srcs, init, batch_size=batch_size)
+    return w
+end
+
+function tmapreduce(f, op, srcs::AbstractArray...;
+                   init=NoInit(),
+                   batch_size= default_batch_size(length(first(srcs)))
+                  )
+    if isempty(first(srcs))
+        if init isa NoInit
```

Comment:
Is this just to throw the same error or is there another reason? If it is for the error then we can throw it straight away.

Reply:
This is a comment on the whole block.

Reply:
Not only throwning the error. Base might do clever magic on empty iterators:
```julia
julia> mapreduce(identity, +, Int[])
0
```